master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,598 @@
+just ship logs; it's too important of a job to mix it with anything else.
+
+---
+
+ship_logs.py: a suitable container entrypoint that ships logs from a child process to the master.
+
+usage: ship_logs.py CMD ARGS...
+
+ship_logs.py will read environment variables set by the master to obtain its configuration.  It
+isn't intended to be useful in any non-managed environments.
+"""
+
+import datetime
+import io
+import json
+import logging
+import os
+import queue
+import re
+import signal
+import ssl
+import subprocess
+import sys
+import threading
+import time
+import traceback
+import urllib.request
+from typing import Any, Dict, IO, Iterator, List, Optional, cast
+
+# Duplicated from determined/__init__.py.  It's nice to keep them in sync.
+LOG_FORMAT = "%(levelname)s: [%(process)s] %(name)s: %(message)s"
+
+
+# Example log message given below.
+# 2022-05-12 16:32:48,757:gc_checkpoints: [rank=0] INFO: Determined checkpoint GC, ...
+# Below regex is used to extract the rank field from the log message.
+# Excluding empty spaces this regex matches rank in the above example as [rank=0]
+rank = re.compile(r"(?P<space1> ?)\[rank=(?P<rank_id>([0-9]+))\](?P<space2> ?)(?P<log>.*)")
+# Below regex is used to extract the message severity from the log message.
+# Excluding empty spaces and delimiter(:) this regex matches message severity level in the above
+# example as INFO
+level = re.compile(
+    r"(?P<space1> ?)(?P<level>(DEBUG|INFO|WARNING|ERROR|CRITICAL)):(?P<space2> ?)(?P<log>.*)"
+)
+lineend = re.compile(rb"[\r\n]")
+
+
+# Interval at which to force a flush.
+SHIPPER_FLUSH_INTERVAL = 1  # How often to make API calls
+
+# Full jitter time on encountering an API exception.
+SHIPPER_FAILURE_BACKOFF_SECONDS = 1
+
+# Max size of the log buffer before forcing a flush.
+LOG_BATCH_MAX_SIZE = 1000
+
+# Max size of the shipping queue before we start to apply backpressure by blocking sends. We would
+# only hit this if we got underwater by three full batches while trying to ship a batch.
+SHIP_QUEUE_MAX_SIZE = 3 * LOG_BATCH_MAX_SIZE
+
+
+def read_newlines_or_carriage_returns(fd: io.RawIOBase) -> Iterator[str]:
+    r"""
+    Read lines, delineated by either '\n' or '\r.
+
+    Unlike the default io.BufferedReader used in subprocess.Popen(bufsize=-1), we read until we
+    encounter either '\n' or \r', and treat that as one line.
+
+    Specifically, io.BufferedReader doesn't handle tqdm progress bar outputs very well; it treats
+    all of the '\r' outputs as one enormous line.
+
+    Args:
+        fd: an unbuffered stdout or stderr from a subprocess.Popen.
+
+    Yields:
+        A series of str, one per line.  Each line always ends with a '\n'.  Each line will be
+        broken to length io.DEFAULT_BUFFER_SIZE, even if the underlying io didn't have a linebreak.
+    """
+    # Ship lines of length of DEFAULT_BUFFER_SIZE, including the terminating newline.
+    limit = io.DEFAULT_BUFFER_SIZE - 1
+    nread = 0
+    chunks = []  # type: List[bytes]
```

Comment:
nit (or maybe clarification for my own self): why the old-style annotation? How old a version of python3 are we accommodating? Do you think it's worth an explicit version check somewhere (as combination documentation for future devs and usefulness for debugging)?

Also re: `chunks`, I should have paid more attention to type (it's a `List[bytes]`, not a `List[ByteString]`), but it took me a while to figure out that this was an accumulated, single line-to-yield rather than a list of them (I guess I read it as lines). I'm not sure what the name says to the reader, so this is just a note to you to think again about the name, and whatever you decide after thinking about it is 👍 to me.

Reply:
I just like the old annotation style I think.  I don't care too much.

To me, "chunks" means "not lines" and "not individual bytes", with no particular meaning to how the chunks were broken up.  I think this is consistent with naming for things like chunked encoding in HTTP; a bunch of chunks make something useful, but there's no particular pattern to how the chunks were broken up.

Reply:
Done.

Thinking more about it, I think the reason I like the old style is putting type annotations in a comment _looks_ like it doesn't affect the runtime.  I _know_ that type annotations inside functions don't affect the runtime behavior, even in some surprising cases, but it doesn't _look_ like that is the case.

Though I agree the new-style annotations are the way of the future, so I can't really complain.  I can stick to this for new code I write.

Reply:
Thanks!

> looks like it doesn't affect the runtime.

I've never thought about this, but I find myself agree with you. That's a nice consequence of the syntax. Though I guess also weird to have comments that are written for a program instead of a human and _do_ affect interpretation (albeit only when type checking). Huh.

Reply:
One more to go:

```shell
% git show 519405d0b79d52ae8d22cd8ea1205db181ece70f | grep ' type: '
```

Reply:
ah, the e2e test, forgot to check there